R/pkg/R/SQLContext.R

```diff
@@ -197,17 +197,40 @@ writeToFileInArrow <- function(fileName, rdf, numPartitions) {
   }
 }
 
-checkTypeRequirementForArrow <- function(dataHead, schema) {
-  # Currenty Arrow optimization does not support raw for now.
-  # Also, it does not support explicit float type set by users. It leads to
-  # incorrect conversion. We will fall back to the path without Arrow optimization.
-  if (any(sapply(dataHead, is.raw))) {
-    stop("Arrow optimization with R DataFrame does not support raw type yet.")
-  }
-  if (inherits(schema, "structType")) {
-    if (any(sapply(schema$fields(), function(x) x$dataType.toString() == "FloatType"))) {
-      stop("Arrow optimization with R DataFrame does not support FloatType type yet.")
+getSchema <- function(schema, firstRow = NULL, rdd = NULL) {
```

Comment:
This had to be a separate function so that Arrow can check the unsupported types from the same `structType` (it partly checked it from the first row before).

We can check the schema early as well but unfortunately the current R schema checking logic is after the data is sent to JVM. I just decided to don't touch existing logic. So ..

The current R path without Arrow:

1. Send the data (R data frame) to JVM
2. Check (and infer if needed) the schema from the first row

R with Arrow:

1. Check (and infer if needed) the schema from the first row
1.1. Fallback if some types of the schema are unsupported.
1.2. Pass
2. Send the data (R data frame) to JVM
3. No-op.

Reply:
I think that's fine